MT29F4G08 NAND Flash Driver

LinderPi opened the pull request:
Introduces a NAND flash driver for Micron's MT29F4G08 memory device that is configurable via devicetree. A generic flash controller driver communicates to the flexible memory controller of STM32 MCUs. Any other NAND flash driver can be implemented to work via FMC in the future. The chip-specific driver provides the flash API for upper layers and enables the on-die ECC feature.

The flash controller is strongly inspired by the STM32 HAL NAND driver following the ONFI specification. However, it optionally uses DMA to transfer a read page. In future, an additional function can be implemented for entire page reads. At the moment this is not required by upper layers and the direct memory transfer is somehow slower to a yet unknown problem.

As it is the first NAND flash driver in Zephyr, some API extensions are foreseen for bad block management by a flash translation layer (FTL). As proposed in #99908, we are currently working on a NFTL disk driver module based on Dhara. A first version is implemented and available [here](https://github.com/endresshauser-lp/dhara/tree/zephyr) for testing.

As an optional enhancement, a bad block map could be kept in memory such that the driver does not need to ask the memory device for bad blocks on every request.

LinderPi committed:
drivers: flash: Add NAND flash driver for MT29F4G08 via STM32 FMC

Introduces a NAND flash driver for Micron's MT29F4G08 memory device that
is configurable via devicetree. A generic flash controller driver
communicates to the flexible memory controller of STM32 MCUs. Any other
NAND flash driver can be implemented to work via FMC in the future. The
chip-specific driver provides the flash API for upper layers and enables
the on-die ECC feature.

As it is the first NAND flash driver in Zephyr, some API extensions are
foreseen for bad block management by a flash translation layer (FTL).

Signed-off-by: Pascal Linder <pascal.linder@zuehlke.com>
Signed-off-by: Tim Pambor <tim.pambor@codewrights.de>

LinderPi committed:
tests: drivers: build_all: flash: Add build test for MT29F4G08

Renames overlays to be board-specific and adds a build test for the newly
introduced NAND flash driver for MT29F4G08 via STM32 FMC.

Signed-off-by: Pascal Linder <pascal.linder@zuehlke.com>

etienne-lms commented on drivers/flash/flash_stm32_fmc_mt29f4g08.c:
```diff
@@ -0,0 +1,336 @@
+			{                                                                          \
+				.write_block_size = DT_INST_PROP(n, page_size),                    \
+				.erase_value = 0xff,                                               \
+			},                                                                         \
```
```suggestion
		.parameters = {                                                                        \
			.write_block_size = DT_INST_PROP(n, page_size),                            \
			.erase_value = 0xff,                                                       \
		},                                                                                  \

LinderPi replied:
Formatter puts it that way. Following change would work:
```
		.parameters = {.write_block_size = DT_INST_PROP(n, page_size),                     \
				       .erase_value = 0xff},                                               \
```

etienne-lms commented on drivers/flash/flash_stm32_fmc_nand.h:
```diff
@@ -0,0 +1,50 @@
+ */
+
+#ifndef __ZEPHYR_DRIVERS_FLASH_STM32_FMC_NAND_H__
+#define __ZEPHYR_DRIVERS_FLASH_STM32_FMC_NAND_H__
```
```suggestion
#ifndef __ZEPHYR_DRIVERS_FLASH_FLASH_STM32_FMC_NAND_H__
#define __ZEPHYR_DRIVERS_FLASH_FLASH_STM32_FMC_NAND_H__
```
to match header file path.

etienne-lms reviewed (commented):
Minor comments only.

LinderPi committed:
drivers: flash: Add NAND flash API extensions

Some API extensions are foreseen for bad block management by a flash
translation layer (FTL).

Signed-off-by: Pascal Linder <pascal.linder@zuehlke.com>

tpambor commented on drivers/flash/Kconfig.stm32_fmc:
```diff
@@ -0,0 +1,17 @@
+	bool "STM32 FMC NAND flash controller"
+	default y
+	depends on DT_HAS_ST_STM32_FMC_NAND_ENABLED
+	select MEMC
```
Furthermore, clock setup is handled by https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_stm32.c driver, which is only compiled if `CONFIG_MEMC=y`, so this is required even if FMC clock rate would not be logged.

LinderPi commented on drivers/flash/flash_stm32_fmc_nand.c:
```diff
@@ -0,0 +1,767 @@
+	(void)FMC_NAND_CommonSpace_Timing_Init(dev_data->instance, &timing,
+					       dev_data->init.NandBank);
+	(void)FMC_NAND_AttributeSpace_Timing_Init(dev_data->instance, &timing,
+						  dev_data->init.NandBank);
```
The STM32 HAL NAND driver does not provide the entire functionality we need (DMA support, status read, etc.). Additionally, there is no intend to further develop it. This is the reasoning I got from ST's official support: "The NAND is not so popular since the xSPI memories we see the FMC usage less and less."

erwango replied:
> The STM32 HAL NAND driver does not provide the entire functionality we need (DMA support, status read, etc.).

Ok, that's clear. However, I need to check internally if `FMC_NAND_()` usage is safe.

etienne-lms commented on drivers/flash/flash_stm32_fmc_mt29f4g08.c:
```diff
@@ -0,0 +1,338 @@
+		*(int *)out =
+			((ret == 0) && (spare_area[0] != 0xFF)) ? NAND_BLOCK_BAD : NAND_BLOCK_GOOD;
+		break;
+	}
```
I don't think these braces (in both `case` sequences)  are useful. There indentation could be confusing. Could you remove them?

etienne-lms commented on drivers/flash/flash_stm32_fmc_nand.c:
```diff
@@ -0,0 +1,767 @@
+		/* Get page data into buffer */
+		for (size_t index = 0; index < dev_data->page_size; index++) {
+			config->page_buffer[index] = sys_read8(NAND_DEVICE);
+		}
```
Using `config->page_buffer` allocated in cache memory is useful when using DMAs, but here, maybe worth the straight load data read into the likely cached `data` buffer.
If using this scheme, you could avoid `flash_stm32_fmc_nand_page_buffer_XXX[]` allocations.

LinderPi commented on drivers/flash/flash_stm32_fmc_mt29f4g08.c:
```diff
@@ -0,0 +1,338 @@
+		ret = flash_stm32_fmc_nand_read_spare_area(config->controller, &address,
+							   spare_area);
+		*(int *)out =
+			((ret == 0) && (spare_area[0] != 0xFF)) ? NAND_BLOCK_BAD : NAND_BLOCK_GOOD;
```
But then I have to set the output value to bad in the else case.

LinderPi commented on drivers/flash/flash_stm32_fmc_nand.c:
```diff
@@ -0,0 +1,755 @@
+	};                                                                                         \
+                                                                                                   \
+	static struct flash_stm32_fmc_nand_data flash_stm32_fmc_nand_data_##n = {                  \
+		FMC_NAND_DMA_CHANNEL(DT_DRV_INST(n), tx_rx)};                                      \
```
Formatter puts it that way.

LinderPi commented on drivers/flash/flash_stm32_fmc_nand.c:
```diff
@@ -0,0 +1,767 @@
+			dev_data->state = NAND_STATE_ERROR;
+			k_sem_give(&dev_data->lock);
+			return -EIO;
+		}
```
I changed it to use Zephyr DMA API only. Different implementations led to various performance results for reading an entire page:
- Initial: 9.5 MiBps
- Busy waiting using `dma_get_status()`: 8.7 MiBps
- Semaphore synchronisation for first transfer: 8.4 MiBps
- Semaphore synchronisation for both transfers: 6.9 MiBps

I decided to go for a semaphore synchronisation after getting the page data into the buffer. It is a good trade-off between a clean solution and good performance.

ttmut commented on drivers/flash/flash_stm32_fmc_mt29f4g08.c:
```diff
@@ -0,0 +1,356 @@
+			flash_mt29f4g08_calculate_address(config, offset);
+
+		ret = flash_stm32_fmc_nand_read_page_chunk(controller, &address, page_offset, chunk,
+							   (uint8_t *)data);
```
Sorry if this was asked before but should not this driver be controller-independent instead? It seems every vendor will have to maintain their own variant of `flash_vendor_mt29f4g08.c`.

tpambor replied:
> I did not actually mean that there should be single driver that covers all NAND chips. I was asking if it is possible to move the calls to STM32 FMC inside `flash_stm32_fmc_mt29f4g08.c` behind some common API so that it becomes vendor-agnostic, such as `flash_mt29f4g08.c`. Otherwise, we are going to end up with `flash_vendorA_mt29f4g08.c`, `flash_vendorB_mt29f4g08.c` whenever someone decides to add support for their controller.

Yes, I wanted to make clear both decisions, why it is flash-chip dependent and why it is flash-controller dependent as others might have the same question. The second part was about the flash-controller dependency.

> > The reason these drivers are NAND controller‑dependent is timing. Parallel NAND flashes have fairly strict and sometimes subtle timing requirements (setup/hold times, access times, etc.), and each controller exposes its own mechanism and parameters for generating those timings. In practice, it is very difficult to define a standardized, controller‑independent interface that maps cleanly onto the timing configuration model of every NAND controller. For the same reason, we also have vendor‑specific bindings for e.g. SDRAM, where the timing parameters are inherently tied to the memory controller’s register layout and capabilities. See, e.g.:
> > https://github.com/zephyrproject-rtos/zephyr/blob/32823f39fbf2663fa1bbb75e60e5743cd6d91579/dts/bindings/memory-controllers/st%2Cstm32-fmc-sdram.yaml#L142-L166

tpambor commented on dts/bindings/memory-controllers/st,stm32-fmc-nand.yaml:
```diff
@@ -0,0 +1,145 @@
+  description: NAND bank.
+
+  properties:
+    reg:
```
The other fmc bindings in tree use `reg`.
https://github.com/zephyrproject-rtos/zephyr/blob/a0fdebcc908ad484c29092c30be5513b9d346651/dts/bindings/memory-controllers/st%2Cstm32-fmc-nor-psram.yaml#L86-L93

https://github.com/zephyrproject-rtos/zephyr/blob/a0fdebcc908ad484c29092c30be5513b9d346651/dts/bindings/memory-controllers/st%2Cstm32-fmc-sdram.yaml#L115-L121

So, maybe keep it to be consistent with the others?

JarmouniA replied:
Makes sense.

SuperHeroAbner commented on include/zephyr/drivers/flash/nand_flash.h:
```diff
@@ -0,0 +1,42 @@
+
+	/* Feature data. */
+	uint8_t feature_data[4];
+};
```
Is this header common for both serial/parallel nand? This feature structure looks for parallel nand with ONFI definition.

tpambor commented:
Currently, no capacity to work on this.
